Add Search

bosschaert opened the pull request:
Contributes to #3

Test URLs:
- Before: https://main--sunstar-engineering--hlxsites.hlx.page/
- After: https://issue-4--sunstar-engineering--hlxsites.hlx.page/

bosschaert commented:
The commits up to now only provide the search box styling.

bosschaert commented:
Marking this issue as ready. It only has the search widget on the screen. I'm planning to add the actual functionality in a different PR.

jindaliiita commented on blocks/header/header.js:
```diff
@@ -20,6 +20,28 @@ function decorateBottomNav() {
+  // TODO specify the correct language in the 'lang' input
+  // TODO specify the correct language in the oninvalid property
+  return `
+    <form method="get" class="search" id="searchform" action="https://www.sunstar-engineering.com/">
```
why action = https://www.sunstar-engineering.com/
and what is the need of id ??

jindaliiita commented on blocks/header/header.css:
```diff
@@ -116,3 +116,58 @@ header nav.nav-top>ul>li>a:first-of-type {
+}
+
+header form.search input.search-text::placeholder {
+  color: rgba(60 70 80 / 50%);
```
i am not sure but i think using hex value for color would be good

bosschaert commented:
Now it contains working search functionality. An example search result is this: : https://issue-4--sunstar-engineering--hlxsites.hlx.page/search?s=adhes

jindaliiita commented on blocks/header/header.css:
```diff
@@ -116,3 +116,58 @@ header nav.nav-top>ul>li>a:first-of-type {
+  width: 14px;
+}
+
+@media (min-width: 77rem) {
```
May be we should start using **width >=**
You can check below comment
https://github.com/hlxsites/sunstar-engineering/pull/37#discussion_r1209974851

bosschaert replied:
Done.

Didn't know you could do this. Is there a difference between `min-width: 77rem` and `width >= 77rem` ?

jindaliiita commented on blocks/header/header.css:
```diff
@@ -116,3 +116,58 @@ header nav.nav-top>ul>li>a:first-of-type {
+header form.search .search-icon {
+  background-color: var(--background-color);
+  background-image: url('/icons/search.svg');
+  background-repeat: no-repeat;
```
background-repeat: no-repeat;
background-position: center;

Some properties i think common for all icons. For eg. - Carousel. If that is the case we can move those in styles.css
https://github.com/hlxsites/sunstar-engineering/blob/main/blocks/carousel/carousel.css#L64-L66

bosschaert replied:
Done. I didn't add them to the carousel yet, but they're in styles.css now.

jindaliiita commented on blocks/header/header.js:
```diff
@@ -20,6 +21,13 @@ function decorateBottomNav() {
 
+function insertSearchWidget() {
+  // Replace the <p>Search</p> text with the search box
+  const searchP = Array.from(document.querySelectorAll('p'))
```
is there any other way rather than making a check on search text content ?

bosschaert replied:
I'd like to replace in the header worddoc where it states the word 'Search' with the search widget, so it's location could be tweaked via the worddoc. However you are right that this is too general. 

I've updated the query to be more specific, hope that works for you.

jindaliiita commented on styles/styles.css:
```diff
@@ -85,6 +87,7 @@ body {
   color: var(--text-color);
   background-color: var(--background-color);
   display: none;
+  -webkit-font-smoothing: antialiased;
```
its already present may be u can take merge

jindaliiita commented on scripts/common.js:
```diff
@@ -0,0 +1,27 @@
+function getSearchWidgetHTML(initialVal) {
```
what exactly is the use of this file when we have scripts.js and lib-franklin.js

bosschaert replied:
You are right, we might just as well put these functions in scripts.js

jindaliiita commented on blocks/search-results/search-results.js:
```diff
@@ -0,0 +1,57 @@
+}
+
+async function searchPages(term) {
+  const resp = await fetch(`${window.location.origin}/query-index.json`);
```
can this function be used here ?
https://github.com/hlxsites/sunstar-engineering/blob/main/scripts/scripts.js#L136

bosschaert replied:
Sounds good - it seems that it will also help me with the paging, which I still need to do. That is #41 so would it be ok with you if I handle this when I look at the paging? At that point the code in scripts.js probably won't look at your _drafts folder any more too.

bosschaert replied:
Done: https://github.com/hlxsites/sunstar-engineering/issues/41#issuecomment-1570187505

jindaliiita commented on blocks/search-results/search-results.js:
```diff
@@ -0,0 +1,57 @@
+  const div = document.createElement('div');
+
+  const summary = document.createElement('h3');
+  summary.className = 'search-summary';
```
classlist.add would be good option

jindaliiita commented on blocks/search-results/search-results.js:
```diff
@@ -0,0 +1,57 @@
+
+  firstPage.forEach((line) => {
+    const res = document.createElement('div');
+    res.className = 'search-result';
```
classlist.add

bosschaert commented:
I see that there are some a11y issues in the search page. I'd like to address them in a different PR.

sdmcraft commented on blocks/search-results/search-results.js:
```diff
@@ -0,0 +1,57 @@
+
+async function searchPages(term) {
+  const resp = await fetch(`${window.location.origin}/query-index.json`);
+  const json = await resp.json();
```
Probably we need a `resp.ok` check here. But based on prior conversation, as you plan to move to the method which already does this in `scripts.js` , it should be ok.

bosschaert replied:
Right yes, that's the plan.

sdmcraft commented on blocks/search-results/search-results.js:
```diff
@@ -0,0 +1,57 @@
+  const json = await resp.json();
+
+  const result = json.data
+    .filter((entry) => (entry.description + entry.title).toLowerCase()
```
```suggestion
    .filter((entry) => (entry.description + ' ' +  entry.title).toLowerCase()
```

bosschaert replied:
Good catch!

bosschaert commented on blocks/header/header.css:
```diff
@@ -117,3 +117,55 @@ header nav.nav-top>ul>li>a:first-of-type {
+  position: absolute;
+  bottom: 8px;
+  right: 0;
+  height: 14px;
+  width: 14px;
```
That would make the icons 20px x 20px which is bigger than on the original... Unless we change the global to 14px x 14px for icons?

sdmcraft replied:
ok yes that 20px X 20px was to match the size. I am thinking of making all icons as background images in header and then perhaps we'll change  the 20px to 14px . But till then, we can leave this as it is.

bosschaert commented:
@sdmcraft my merge from main commit also hopefully addresses the issues you mentioned.

bosschaert commented on blocks/header/header.css:
```diff
@@ -125,4 +125,59 @@ header nav.nav-top>ul>li>a:first-of-type {
+  right: 0;
+  height: 14px;
+  width: 14px;
+  min-width: 14px;
```
@sdmcraft I had to add this `min-width` as it was inhering a new value of `160` from button in styles.css, was that the intention?

sdmcraft replied:
ok yes that's the global button style. Probably you can override this style and specify  `button` in the selector.

sdmcraft reviewed (commented):
The margin/padding etc. seem to be missing from the mobile view. Also the font also looks a bit larger

Franklin
<img width="662" alt="image" src="https://github.com/hlxsites/sunstar-engineering/assets/1191451/27687ed9-615c-4ea6-b880-ff1dfcdb24e7">

Original
<img width="629" alt="image" src="https://github.com/hlxsites/sunstar-engineering/assets/1191451/8af02a3f-d074-4bee-a2ca-6ed5aa3fbdac">

sdmcraft reviewed (commented):
The margin/padding etc. seem to be missing from the mobile view. Also the font also looks a bit larger

Franklin
<img width="662" alt="image" src="https://github.com/hlxsites/sunstar-engineering/assets/1191451/27687ed9-615c-4ea6-b880-ff1dfcdb24e7">

Original
<img width="629" alt="image" src="https://github.com/hlxsites/sunstar-engineering/assets/1191451/8af02a3f-d074-4bee-a2ca-6ed5aa3fbdac">

sdmcraft commented:
There's this greyed out text which seems to be category or topic. Perhaps something to figure out separately.
![image](https://github.com/hlxsites/sunstar-engineering/assets/1191451/aaff8687-130f-44b3-9b71-b4e695f6a380)

bosschaert commented:
Hi @sdmcraft,

> The margin/padding etc. seem to be missing from the mobile view. Also the font also looks a bit larger

I adjusted the font sizes. Interestingly the css property `text-size-adjust` in the original site made fonts appear different than exactly specified in mobile views. I applied this property to the Franklin styles.css too.

I created a separate issue to deal with the mobile paragraph margins, which is a general issue for all pages I think: https://github.com/hlxsites/sunstar-engineering/issues/65

> There's this greyed out text which seems to be category or topic. Perhaps something to figure out separately. 

I created https://github.com/hlxsites/sunstar-engineering/issues/66 for this.

Besides the ones that I created issues for, I think I addressed all your comments at this stage.

sdmcraft commented on blocks/header/header.js:
```diff
@@ -43,15 +44,22 @@ export default async function decorate(block) {
       nav.innerHTML = fetchedNav.querySelectorAll(':scope>div')[idx].innerHTML;
       navDecorators[navClass](nav);
+
+      if (navClass === 'nav-bottom') {
+        nav.append(getSearchWidget());
+      }
```
i think this can be moved to `decorateBottomNav`

bosschaert replied:
Ah yes, I missed that possibility. Thanks!

sdmcraft reviewed (approved):
Thanks. Look good. Added 1 minor comment.
